arch/src/aarch64/gic.rs

```diff
@@ -21,104 +18,86 @@ pub enum Error {
 }
 type Result<T> = result::Result<T, Error>;
 
-/// Create a GICv3 device.
+/// Trait for GIC devices.
+pub trait GICDevice {
+    /// Returns the GIC version of the device
+    fn version(&self) -> u32;
+
+    /// Returns the file descriptor of the GIC device
+    fn device_fd(&self) -> &DeviceFd;
+
+    /// Returns an array with GIC device properties
+    fn device_properties(&self) -> &[u64];
+}
+
+/// Create a GIC device.
 ///
-/// Logic from this function is based on virt/kvm/arm/vgic/vgic-kvm-device.c from linux kernel.
-pub fn create_gicv3(vm: &VmFd, vcpu_count: u8) -> Result<DeviceFd> {
-    /* We are creating a V3 GIC.
-     As per https://static.docs.arm.com/dai0492/b/GICv3_Software_Overview_Official_Release_B.pdf,
-     section 3.5 Programmers' model, the register interface of a GICv3 interrupt controller is split
-     into three groups: distributor, redistributor, CPU.
-     As per Figure 9 from same section, there is 1 Distributor and multiple redistributors (one per
-     each CPU).
-    */
+/// It will try to create by default a GICv3 device. If that fails it will try
+/// to fall-back to a GICv2 device.
+pub fn create_gic(vm: &VmFd, vcpu_count: u64) -> Result<Box<GICDevice>> {
+    GICv3::new(vm, vcpu_count).or_else(|_| GICv2::new(vm))
+}
+
+/// Initialize a GIC device
+pub fn init_device(vm: &VmFd, gic_device_type: u32) -> Result<DeviceFd> {
```

Comment:
hmmm, how can `init_device` (and `set_device_attribute`, `finalize_device`, for that matter) be parts of the Trait object? They are used to create the object itself? How can we call the Trait method on an object that has not yet been constructed?

Reply:
If the method is static you can call it even if the object hasn't been instantiated yet. For example something like this would work:

```
pub trait GICDevice {
    fn version() -> u32;

    fn init_device() -> bool {
        Self::version() == 1
    }
}

pub struct GICv2 {
    initialized: bool
}

impl GICDevice for GICv2 {
    fn version() -> u32 {
        1
    }

}

impl GICv2 {
    fn new() -> GICv2 {
        GICv2 {
            initialized: Self::init_device()
        }
    }
}
```

Reply:
oh, great! I didn't know that. Thanks a lot.

Reply:
Hmmm, I tried this out, but when I declare static Trait methods the compiler complains that the trait cannot be made into an object (meaning by the `Box<GICDevice>` I'm returning from the `new` method.

I have been looking around and it seems that there is no nice way around this issue. Please let me know, if I'm msising something

Reply:
Ok, I dealt with it by marking the static methods of the Trait with `where Self: Sized`, e.g.:

```rust
pub trait GICDevice {
    fn init_device() -> bool where Self: Sized {
        true
    }
}
```
essentially, saying that these will not be called on the object itself. Does that make sense to you?

Reply:
Nice. Sounds good. It probably makes sense, but I have to see the entire commit anyway.